ufl/algorithms/strip_form_argument_data.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+"""Algorithm for replacing form arguments with 'stripped' versions where any
+data-carrying objects have been extracted to a mapping."""
+
+from ufl.classes import Form
+from ufl.classes import Argument, Coefficient, Constant
+from ufl.classes import FunctionSpace, TensorProductFunctionSpace, MixedFunctionSpace
+from ufl.classes import Mesh, MeshView, TensorProductMesh
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.replace import Replacer
+from ufl.corealg.multifunction import MultiFunction
+
+
+class DataStripper(MultiFunction):
+    def __init__(self):
+        super().__init__()
+        self.mapping = {}
+
+    def argument(self, o):
+        o_new = Argument(_strip_function_space(o.ufl_function_space()),
+                         o.number(), o.part())
+        return self.mapping.setdefault(o, o_new)
+
+    def coefficient(self, o):
+        o_new = Coefficient(_strip_function_space(o.ufl_function_space()),
+                            o.count())
+        return self.mapping.setdefault(o, o_new)
+
+    def constant(self, o):
+        o_new = Constant(_strip_domain(o.ufl_domain()), o.ufl_shape,
+                         o.count())
+        return self.mapping.setdefault(o, o_new)
+
+    expr = MultiFunction.reuse_if_untouched
+
+
+def strip_form_argument_data(form):
+    """Convert an UFL expression to a new UFL expression where the form
+    arguments have been replaced with pure UFL copies. The new expression is
+    returned as well as a mapping from the new arguments to the old ones.
+
+    This function is useful for forms containing augmented UFL objects that
+    hold references to large data structures. These objects are be extracted
+    into the mapping allowing the form to be cached without leaking memory.
+    """
+    if not isinstance(form, Form):
+        raise ValueError(f"{form} must be a Form")
+
+    data_stripper = DataStripper()
+    stripped_form = map_integrand_dags(data_stripper, form,
+                                       reconstruct_integral_domain=True)
+    # We invert the mapping so it can be passed straight into ufl.replace
+    return stripped_form, {v: k for k, v in data_stripper.mapping.items()}
+
+
+def attach_form_argument_data(form, mapping):
+    """Convert an UFL expression to a new one where the form arguments have been
+    replaced by those in the provided mapping. The keys of the map correspond to
+    the objects to be replaced and the values are the objects they are
+    to be replaced with.
+
+    Note that this function is simply the opposite of strip_form_arguments.
+    """
+    if not isinstance(form, Form):
+        raise ValueError(f"{form} must be a Form")
+
+    return map_integrand_dags(Replacer(mapping), form,
+                              reconstruct_integral_domain=True)
```

Comment:
What is the difference between this function and `replace`?

Reply:
The `reconstruct_integral_domain` argument. I didn't want to add a keyword argument to `replace`.

Reply:
But if the `test_mixed_function_space` error does turn out to be a bug then I can abandon `reconstruct_integral_domain` and this function can go.